gpcontrib/arenadata_toolkit/src/drops_track.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Track unlink hook events.
+ */
+
+#include "drops_track.h"
+
+#include "lib/ilist.h"
+#include "storage/lwlock.h"
+#include "storage/ipc.h"
+#include "storage/shmem.h"
+
+#include "arenadata_toolkit_guc.h"
+
+#define TRACK_NODE_GET(track, i) (void *)(track->nodes + i * sizeof(drops_track_node_t));
+
+/*
+ * Drop track element. Stores just relfilenode
+ * and dbid.
+ */
+typedef struct
+{
+	Oid			relNode;
+	Oid			dbNode;
+}	track_relfilenode_t;
+
+/* Doubly linked list node of dropped file nodes */
+typedef struct
+{
+	dlist_node	node;
+	track_relfilenode_t relfileNode;
+}	drops_track_node_t;
+
+/* Drops track */
+typedef struct
+{
+	dlist_head	head;
+	uint32_t	used_count;		/* count of used nodes */
+	int			unused_idx;		/* next unused idx or -1 if unknown; for
+								 * faster search */
+	char		nodes[FLEXIBLE_ARRAY_MEMBER];	/* array of drops_track_node_t */
+}	drops_track_t;
+
+static shmem_startup_hook_type next_shmem_startup_hook = NULL;
+static drops_track_t *drops_track;
+LWLock	   *drops_track_lock;
+
+static Size
+drops_track_calc_size()
+{
+	Size		size;
+
+	size = offsetof(drops_track_t, nodes);
+	size = add_size(size, mul_size(drops_count, sizeof(drops_track_node_t)));
+
+	return size;
+}
+
+static void
+drops_track_hook(void)
+{
+	bool		found;
+	Size		size = drops_track_calc_size();
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	drops_track = ShmemInitStruct("adb_track_files_drops", size, &found);
+
+	if (!found)
+	{
+		drops_track->used_count = 0;
+		drops_track->unused_idx = 0;
+		dlist_init(&drops_track->head);
+
+		for (uint32_t i = 0; i < drops_count; i++)
+		{
+			drops_track_node_t *track_node = TRACK_NODE_GET(drops_track, i);
+
+			track_node->relfileNode.relNode = InvalidOid;
+			track_node->relfileNode.dbNode = InvalidOid;
+		}
+	}
+
+	drops_track_lock = LWLockAssign();
+
+	LWLockRelease(AddinShmemInitLock);
+
+	if (next_shmem_startup_hook)
+		next_shmem_startup_hook();
+}
+
+void
+drops_track_init(void)
+{
+	RequestAddinLWLocks(1);
+	RequestAddinShmemSpace(drops_track_calc_size());
+
+	next_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = drops_track_hook;
+}
+
+void
+drops_track_deinit(void)
+{
+	shmem_startup_hook = next_shmem_startup_hook;
+}
+
+/* Find unused node in linked list. */
+static drops_track_node_t *
+find_empty_node()
+{
+	drops_track_node_t *track_node = NULL;
+
+	if (drops_track->unused_idx >= 0)
+	{
+		track_node = TRACK_NODE_GET(drops_track, drops_track->unused_idx);
+		drops_track->unused_idx++;
+		if (drops_track->unused_idx >= drops_count)
+			drops_track->unused_idx = -1;
+		else
+		{
+			drops_track_node_t *unused_node = TRACK_NODE_GET(drops_track, drops_track->unused_idx);
+
+			if (unused_node->relfileNode.relNode != InvalidOid)
+				drops_track->unused_idx = -1;
+		}
+	}
+	else
+	{
+		for (uint32_t i = 0; i < drops_count; i++)
```

Comment:
It looks that once we set `drops_track->unused_idx = -1`, we'll always need to do linear search for an empty node. For a long lasting instance of GPDB we'll be here anyway sooner or later. Maybe, instead of doing this linear search, add all nodes, that can be recycled after `drops_track_move()`, into a separate list of empty nodes. Push and pop from this list can be made constant complexity. What do you think?